src/wp-admin/includes/post.php

```diff
@@ -1440,7 +1440,7 @@ function get_sample_permalink( $post, $title = null, $name = null ) {
 	$original_filter = $post->filter;
 
 	// Hack: get_permalink() would return plain permalink for drafts, so we will fake that our post is published.
-	if ( in_array( $post->post_status, array( 'draft', 'pending', 'future' ), true ) ) {
+	if ( in_array( $post->post_status, array( 'auto-draft', 'draft', 'pending', 'future' ), true ) ) {
```

Comment:
LGTM.

Reply:
> the second part of the fix is to prevent get_post_status from calling get_post
> ...
> I'm not entirely certain about all the consequences here

Yep, thinking this is not such a good idea. It might result is some "unexplained" regressions in other unexpected code/places as (as far as I see) that deals with post caching.

Permalinks for non-published posts are a tricky thing. They are considered "temporary", and the users can edit them. Generally auto-draft posts should never have a permalink as these posts "don't exist" in the real sense. They are simply placeholders indicating a possible intention to maybe have a post with that ID :)

Reply:
> Perhaps another option here is to ensure the post is saved as draft before trying to get or set a permalink for it?

I think that probably has bigger downsides as the user may leave the editor right away before saving or typing or anything and over time, we'll just end up with a huge numbers of draft posts. I believe that's why auto-drafts exist in the first place.

Reply:
> the user may leave the editor right away before saving or typing or anything and over time, we'll just end up with a huge numbers of draft posts

Yes, it's possible. However if we have to set a permalink that means the user at least has typed a post title, i.e. they have started a new post (otherwise how would WP know what to use to make a permalink). Thinking that having post title is enough of a reason to save a draft. 

If a lot of these posts accumulate over time, they will all show in the Posts list table anyways, and the users will be able to either continue them or delete them.

Reply:
And why do we have a different behavior for pages? For pages, it's possible to set the slugs right away and this functions returns the right permalink right away. This is something I was not able to track down.

Reply:
> For pages, it's possible to set the slugs right away and this functions returns the right permalink right away.

Right, looking at this again best imho would be to save as a draft if it is an auto-draft at that point, then do that "permalink hack". I.e. the change here will be that a new draft post is created when someone edits the permalink before doing anything else like adding title, content, etc.

Just adding `'auto-draft'` to that "hack" (i.e. allowing auto-drafts to have proper permalinks) may open some edge cases like permalinks for non-existing post IDs as old auto-drafts are deleted without warning (garbage collection).

Reply:
> Right, looking at this again best imho would be to save as a draft if it is an auto-draft at that point, then do that "permalink hack". I.e. the change here will be that a new draft post is created when someone edits the permalink before doing anything else like adding title, content, etc.

Unfortunately, that's not possible. `permalink_template` in the REST API response is what indicates to the site editor that the current post type and site actually supports pretty permalink, so if this property is not defined the UI to show the permalink input is not shown at all. Is there any other way (using the REST API) to check whether the current post supports pretty permalinks?

Reply:
I took the liberty of pushing changes to `get_post_status` that should ensure that normalisation is only skipped when intended, by looking at the `WP_Post`'s `filter` instance variable. This follows the precedent set in `get_permalink` and `wp_force_plain_post_permalink`.

Does this assuage others' concerns too, or should other measures be taken?

Reply:
> pushing changes to get_post_status that should ensure that normalisation is only skipped when intended

@mcsf Yea, looks like this removes most of the previous concerns about caching/refreshing the caches. As far as I see normalization can be skipped only for auto-drafts. This _shouldn't_ affect anything since auto-drafts are by definition "empty" or "pre-determined content" placeholders. On the other hand looking at the sheer number of (plugins) code that "does something" with WP_Post, there might be unexpected edge cases even with these restrictions, so perhaps some inline docs there would be useful.

> permalink_template in the REST API response is what indicates to the site editor that the current post type and site actually supports pretty permalink

@youknowriad if that's the main problem here, why not fix it in the REST API instead of changing the `get_sample_permalink()` function? Maybe even there can be another property in the REST response indicating that the post type supports permalinks? Then it would be very easy to control the UI from PHP. I realize this is a different direction after all this time, but do you think it may work better and eventually be useful in the future?

Reply:
I think there are two questions in one here:

 1- I think the permalink_template value is broken in the REST API for auto-drafts. It's inconsistent with other post types for no valid reason and this is how it's computed https://github.com/WordPress/wordpress-develop/blob/4304be07f3042184dfb5c0390b83a4d05a01a43a/src/wp-includes/rest-api/endpoints/class-wp-rest-posts-controller.php#L1991-L1995 A call to `get_sample_permalink`. For me it's logical to fix it in `get_sample_permalink` but it's actually the purpose of this function. Now, if you think there's another way to solve this in the REST API, I'm all for it. REST API for me are just proxies to WP functions. 
 
 2- I do think post types endpoints lack a lot of "supports" keys. We face this a lot in Gutenberg and we end up often having to hard code post types. Things like:

 - Whether the post type support permalinks.
 - Whether the post type support auto-saves.

But there are a lot more and It would be great to add these things. It seems like a big effort.

Reply:
> For me it's logical to fix it in get_sample_permalink()...

Right, makes sense. 

> if you think there's another way to solve this in the REST API

Was thinking the REST API can return just enough data (hard-coded?) to trigger the proper UI when the post is an auto-draft. However that will not run the `get_sample_permalink` filter. Seems running it may be desirable. Returning minimal data from `get_sample_permalink()` would work too, also in cases outside the REST API.

May be overthinking this but as mentioned above for maximum backwards compatibility may be good to add another param to get_sample_permalink(). Perhaps something like:

`get_sample_permalink( $post, $title = null, $name = null, $for_auto_draft = false )`

Then use that in WP_REST_Posts_Controller:

`$sample_permalink = get_sample_permalink( $post->ID, $post->post_title, '', true );`

Reply:
> post types endpoints lack a lot of "supports" keys. We face this a lot in Gutenberg

Thinking I've seen a discussion about outputting all the `post_type_supports` settings for a particular post type? At first look this doesn't seem hard to do, just "translating" the relevant part of the `$_wp_post_type_features` PHP global and adding it to a REST response. Guessing this has been tried before and there were problems? I'll be glad to help do that or resolve any problems if I can :)

Reply:
> May be overthinking this but as mentioned above for maximum backwards compatibility may be good to add another param to get_sample_permalink(). Perhaps something like:

I'm happy to add this, I don't mind it but I'm just noting that this comment already exists in the function. 

> 	// Hack: get_permalink() would return plain permalink for drafts, so we will fake that our post is published.

In other words, we don't have a `$for_draft` argument (or `$for_pending` in these cases). In other words, I wonder if "auto-drafts" were just forgotten in the first place in that "hack".

Reply:
I think we're overthinking a bit. We should obviously be careful about back-compat and with areas of undefined behaviour that could affect users' edge cases out there... but IMO introducing `$for_draft` is going a bit far. Like Riad, I'm more inclined to think that auto-drafts were forgotten, and that fixing that shouldn't really affect anyone.

Reply:
> think that auto-drafts were forgotten

Not exactly. The logic/design intent for auto-drafts is that they are not considered "posts" as the users have not provided any content for them yet. They are placeholders for future posts (that WP eventually auto-deletes) just to get the self-incrementing table row ID (the post_ID) from the DB. 

However this doesn't (shouldn't) affect the decision here. The only thing that matters is to keep back-compat as best as possible. Adding `auto-draft` to the post statuses that support sample permalinks changes the current behavior of `get_sample_permalink()`. I agree the change is very very small, but it's not impossible to cause some sort of issue somewhere. 

Generally the easiest thing to do in such cases is to make this a new "special case" and add some way to detect it in `get_sample_permalink()`. Adding another param seems to fit fest? 

So I kind of prefer if there is another param despite that the chance for a back-compat problem is super small. (Perhaps this can also be seen as an example of how to handle similar cases in the future.)

Reply:
> just noting that this comment already exists in the function...

Yea, I agree, not the best way to do this, by far (pretend that the post is published while calculating a permalink). But this has been in core for so many years, all the quirks this code may be causing are not a part of how it works, and core and plugins have learned how to deal with them and are expecting them. 

This is, imho, one of the hardest things to account for when working with an old codebase :)